src/core/ledger/grainAllocation.js

```diff
@@ -51,7 +60,7 @@ export type Special = "SPECIAL";
 export type AllocationPolicy = RegularPolicy | SpecialPolicy;
 
 export type RegularPolicy = {|
-  +policyType: Immediate | Balanced,
+  +policyType: Immediate | Recent | Balanced,
   +budget: G.Grain,
 |};
```

Comment:
A `RegularPolicy` is characterized by the fact that it only has a budget. However, the Recent policy needs a discount factor, so it actually can't be a regular policy. Suggest making:

```js

export type RecentPolicy = {|
  +policyType: "RECENT",
  +budget: G.Grain,
  +discountFactor: number,
|}
```

However, if we do this, then we'll have a strange `AllocationPolicy` line:

```js
export type AllocationPolicy = RegularPolicy | SpecialPolicy | RecentPolicy;
```

I recommend having a separate commit (and separate PR) which renames the policies and splits `RegularPolicy` into just `ImmediatePolicy` and `BalancedPolicy`. Could do that one first, which would follow our [recommended](https://github.com/sourcecred/sourcecred/blob/master/CONTRIBUTING.md) workflow.

Reply:
What if `AllocationPolicy` was a type with some `policyType`, `budget`, and `receipts` function?  This way, whenever we add or rename policies, we don't have to redefine the `AllocationPolicy`.  This feels like the essence of an `AllocationPolicy`.

```js
// each policy has a receipts function of the same type, so let's define that type
export type ReceiptGenerator = (
   policy: AllocationPolicy,
   identities: ProcessedIdentities
) => $ReadOnlyArray<GrainReceipt>;

export type AllocationPolicy = {
   +policyType: string,
   +budget: G.Grain,
   +receipts: ReceiptGenerator
}
```

It also lets us remove the `receipts` function and instead do call the `AllocationPolicy`'s own `receipts` function:
```js
export function computeAllocation(
  policy: AllocationPolicy,
  identities: $ReadOnlyArray<AllocationIdentity>
): Allocation {
  const validatedPolicy = _validatePolicy(policy);
  const processedIdentities = _processIdentities(identities);
  return _validateAllocationBudget({
    policy,
    receipts: validatedPolicy.receipts(processedIdentities), // <-- this 
    id: randomUuid(),
  });
}

// REMOVE
function receipts(
  policy: AllocationPolicy,
  identities: ProcessedIdentities
): $ReadOnlyArray<GrainReceipt> {
  switch (policy.policyType) {
    case "IMMEDIATE":
      return immediateReceipts(policy.budget, identities);
    case "BALANCED":
      return balancedReceipts(policy.budget, identities);
    case "SPECIAL":
      return specialReceipts(policy, identities);
    // istanbul ignore next: unreachable per Flow
    default:
      throw new Error(`Unknown policyType: ${(policy.policyType: empty)}`);
  }
}
```

Reply:
AllocationPolicy needs to be serializable:

Allocation has the policy:
https://github.com/sourcecred/sourcecred/blob/db854f6925e15601428073641da3484b8d1b3241/src/core/ledger/grainAllocation.js#L70-L73

Allocation is included in a distribution:
https://github.com/sourcecred/sourcecred/blob/db854f6925e15601428073641da3484b8d1b3241/src/core/ledger/distribution.js#L8-L12

And a distribution is included in the ledger:
https://github.com/sourcecred/sourcecred/blob/db854f6925e15601428073641da3484b8d1b3241/src/core/ledger/ledger.js#L776-L779

Since a function isn't serializable, this wouldn't work. Nice suggestion tho.

Reply:
Just to make sure I understand:
```js
export type RegularPolicy = BalancedPolicy | ImmediatePolicy;

export type BalancedPolicy = {|
  +policyType: Balanced,
  +budget: G.Grain,
|}

export type ImmediatePolicy = {|
  +policyType: Immediate,
  +budget: G.Grain,
|}
```
I'm a fan of this style where each policy has it's own `______Policy` object and parser, since we can compose them into groups like `RegularPolicy`.  If we're on the same page, I'll open a separate PR.